### PR TITLE
Delete StatefulTransform and move logic into AMDAIE passes

### PR DESCRIPTION
**Pull request by jtuyls**
This PR removes `AMDAIEStatefulTransform` and `AMDAIEAssignLockIDs` by moving logic into `AMDAIEBufferization`, `AMDAIEAcquireReleaseToUseLock` and `AMDAIELowerToAIE`. This gets rid of an intermediate layer of logic operating on `aie.objectfifo` by going from logical objectFifos directly to aie.buffer. This reduces the overall amount of code/complexity needed to get to the same result.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+      logicalObjFifoToIndex[logicalObjFifo] = 0;
+    size_t bufferIndex = logicalObjFifoToIndex[logicalObjFifo] % buffers.size();
+    for (Operation *userOp : acqOp->getUsers()) {
+      auto accessOp = dyn_cast<AMDAIE::LogicalObjectFifoAccessOp>(userOp);
+      if (!accessOp) {
+        acqOp.emitOpError() << "currently only supports "
+                               "`amdaie.logicalobjectfifo.access` users";
+        return WalkResult::interrupt();
+      }
+      auto bufferOp = buffers[bufferIndex];
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:145]: Expand `auto`.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    FailureOr<AMDAIE::LockOp> maybeLockOp =
+        getLockToBeUsed(maybeLogicalObjFifo.value(), tileOp, relOp.getPort(),
+                        LockAction::Release);
+    if (failed(maybeLockOp)) return WalkResult::interrupt();
+
+    rewriter.setInsertionPoint(relOp);
+    rewriter.create<AMDAIE::UseLockOp>(relOp.getLoc(), maybeLockOp.value(),
+                                       LockAction::Release, relSize);
+    // rewriter.eraseOp(relOp);
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:184]: Leftover code.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td [hunk 3]
```diff
@@ -17,6 +17,12 @@ def AMDAIEAccessToAcquireRelease :
   let constructor = "mlir::iree_compiler::AMDAIE::createAMDAIEAccessToAcquireReleasePass()";
 }
 
+def AMDAIEAcquireReleaseToUseLock :
+    Pass<"iree-amdaie-acquire-release-to-use-lock", ""> {
+  let summary = "Convert acquire/release synchornization stubs to `amdaie.use_lock`";
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td:22]:
```suggestion
  let summary = "Convert acquire/release synchronization stubs to `amdaie.use_lock`";
```

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/Support/MathExtras.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+
+#define DEBUG_TYPE "iree-amdaie-acquire-release-to-use-lock"
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:17]: Potentially, yes, but maybe it makes sense to go through an intermediary as going to `use_lock` requires a loop unroll as well to be valid + buffers. In any case, I think that's out of scope of this PR.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEAttrs.td [hunk 5]
```diff
@@ -21,6 +21,17 @@ def AMDAIE_CopyOpOperateOn: I32EnumAttr<"CopyOpOperateOn",
 {
 }
 
+def AMDAIE_LockAction: I32EnumAttr<"LockAction",
+  "The action to be used performed on a lock",
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEAttrs.td:25]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td [hunk 6]
```diff
@@ -229,6 +229,33 @@ def AMDAIE_LockOp: AMDAIE_Op<"lock", [
   let assemblyFormat = [{ `(` $tile `(` $value `)` (`,` $init_value^)? `)` attr-dict }];
 }
 
+def AMDAIE_UseLockOp: AMDAIE_Op<"use_lock"> {
+  let summary = "Represents a physical lock on an AIE tile.";
+  let description = [{
+    This operation represents a lock on an AIE tile. The operation is fully 
+    specified through a tile and lock ID value, designating the exact physical
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td:236]: Thanks for catching. I didn't update this ops doc it seems.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:145]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp (hunk 2 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:184]: Removed it.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    IRRewriter rewriter(parentOp->getContext());
+
+    WalkResult res = parentOp->walk([&](AMDAIE::CoreOp coreOp) {
+      // Loops need to be unrolled based on on the depths of the logical
+      // objectFifos so `amdaie.use_lock` ops can be inserted correctly for
+      // double buffering purposes, without need for a dependency on the loop
+      // induction variable.
+      if (failed(coreLoopUnroll(rewriter, coreOp))) {
+        return WalkResult::interrupt();
+      }
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:215]: Why don't we separate this out of this pass ?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 8]
```diff
@@ -33,55 +37,89 @@ using namespace xilinx;
...
+    const std::pair<AIE::LockOp, AIE::LockOp> &locks) {
+  OpBuilder::InsertionGuard g(rewriter);
+  Block &endBlock = memOp->getRegion(0).getBlocks().back();
+  assert(!endBlock.getOps<AIE::EndOp>().empty() &&
+         "expected last block to have aie.end");
+  Block *lastDmaBlock = endBlock.getSinglePredecessor(),
+        *dmaBlock = rewriter.createBlock(&endBlock),
+        *bdBlock = rewriter.createBlock(&endBlock);
+
+  // create DMA channel
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:57]:
```suggestion
  // Create DMA channel.
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 9]
```diff
@@ -692,55 +487,357 @@ LogicalResult controlCodeToAie(IRRewriter &rewriter,
...
+          target.getDefiningOp());
+  if (!targetObjFifoLikeOp) {
+    return connectionOp.emitOpError()
+           << "expected target to be an logical objFifo-like op";
+  }
+  if (targetObjFifoLikeOp.getMemorySpaceAsUInt() == 0) {
+    for (AMDAIE::ChannelOp channel : consumerChannels) {
+      auto tileOp = channel.getTileOp();
+      auto shimDmaAllocOp = rewriter.create<AIE::ShimDMAAllocationOp>(
+          rewriter.getUnknownLoc(), "shim_" + std::to_string(connectionIndex++),
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:678]: There should be a cleaner way to do this for both source and target ?
How about creating a struct which we fill in (eg: `mixedSourceSizes()` vs `mixedTargetSizes()`, `MM2S` vs `S2MM`, etc) before invoking the function ?

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp (hunk 7 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp:215]: This unrolling is very specific to the conversion from acquire/release to use_lock, which would not be valid without the unrolling first.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp (hunk 9 above)
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp]: Oh okay, makes sense. No worries.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 10]
```diff
@@ -90,226 +77,110 @@ AIE::BDDimLayoutArrayAttr convertSizeStrideToBDDimLayoutArrayAttr(
...
-  }
+AIE::ShimDMAAllocationOp AIEDeviceBuilder::createShimDmaAllocation(
+    Block *deviceBlock, AMDAIE::TileOp tileOp, AIE::DMAChannelDir dmaChannelDir,
+    uint8_t channel, MemRefType memrefType, int &connectionIndex) {
+  OpBuilder::InsertionGuard g(rewriter);
+  auto shimDmaAllocOp = rewriter.create<AIE::ShimDMAAllocationOp>(
+      rewriter.getUnknownLoc(), "shim_" + std::to_string(connectionIndex++),
+      dmaChannelDir, channel, getConstantIndexOrAssert(tileOp.getCol()));
+  rewriter.setInsertionPointToStart(deviceBlock);
+  auto symName = shimDmaAllocOp.getSymName();
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 11]
```diff
@@ -90,226 +77,110 @@ AIE::BDDimLayoutArrayAttr convertSizeStrideToBDDimLayoutArrayAttr(
...
+  };
+
+  // create Bd blocks
+  Block *succ = nullptr, *curr = bdBlock;
+  for (size_t blockIndex = 0; blockIndex < bufferOps.size(); ++blockIndex) {
+    if (blockIndex == bufferOps.size() - 1) {
+      succ = bdBlock;
+    } else {
+      succ = rewriter.createBlock(&endBlock);
     }
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:142]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 12]
```diff
@@ -90,226 +77,110 @@ AIE::BDDimLayoutArrayAttr convertSizeStrideToBDDimLayoutArrayAttr(
...
-        dyn_cast<AMDAIE::ChannelOp>(consumerChannel.getDefiningOp());
-    if (!channelOp) {
-      return connectionOp.emitOpError()
-             << "found non-`amdaie.channel` source channel";
+    rewriter.create<AIE::UseLockOp>(rewriter.getUnknownLoc(), relLock,
+                                    AIE::LockAction::Release, relNum);
+    rewriter.create<AIE::NextBDOp>(rewriter.getUnknownLoc(), succ);
+  };
+
+  // create Bd blocks
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp]: Done

**Pull request merged**